"Extends" with plugins displays a confusing error message

**Tell us about your environment**

* **ESLint Version:** 3.16.1
* **Node Version:** 7.6.0
* **yarn Version:** 0.20.3

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**
```js
module.exports = {
  extends: ['airbnb', 'plugin:flowtype/recommended'],
  plugins: ['flowtype'],
  rules: {
    'flowtype/define-flow-type': 1,
    'flowtype/use-flow-type': 1,
    'flowtype/delimiter-dangle': [2, 'always'],
    'flowtype/semi': [2, 'always'],
  },
  settings: {
    flowtype: {
      onlyFilesWithFlowAnnotation: true,
    },
  },
};
```

package.json:
```json
...
"eslint-plugin-flowtype": "2.30.0",
...
```

**What did you expect to happen?**
My code would be linted with 'flowtype' rules.

**What actually happened? Please include the actual, raw output from ESLint.**
I get:
```
Oops! Something went wrong! :(

ESLint couldn't find the config "plugin:flowtype/recommended" to extend from. Please check that the name of the config is correct.
```

@danny-andrews What do you mean by "it works"? 
The report is complaining about the config, when the actual error I get if I revert the commit is "Error: Cannot find module 'babel-eslint'"
Is that what you mean?

@alberto Yes, I guess I thought that was a different error.
To clarify:
On 360dbe4:
  without `babel-eslint` installed: `ESLint couldn't find the config "plugin:flowtype/recommended" to extend from. Please check that the name of the config is correct.`
  with `babel-eslint` installed: Works as expected with no errors.
On Pre-360dbe4:
  without `babel-eslint` installed: `Cannot find module 'babel-eslint'`
  with `babel-eslint` installed: Works as expected with no errors.

--- a/lib/config/config-file.js
+++ b/lib/config/config-file.js
@@ -184,6 +184,22 @@ function loadPackageJSONConfigFile(filePath) {
 }
 
 /**
+ * Creates an error to notify about a missing config to extend from.
+ * @param {string} configName The name of the missing config.
+ * @returns {Error} The error object to throw
+ * @private
+ */
+function configMissingError(configName) {
+    const error = new Error(`Failed to load config "${configName}" to extend from.`);
+
+    error.messageTemplate = "extend-config-missing";
+    error.messageData = {
+        configName
+    };
+    return error;
+}
+
+/**
  * Loads a configuration file regardless of the source. Inspects the file path
  * to determine the correctly way to load the config file.
  * @param {Object} file The path to the configuration.
@@ -199,6 +215,9 @@ function loadConfigFile(file) {
             config = loadJSConfigFile(filePath);
             if (file.configName) {
                 config = config.configs[file.configName];
+                if (!config) {
+                    throw configMissingError(file.configFullName);
+                }
             }
             break;
 
@@ -341,6 +360,33 @@ function getLookupPath(configFilePath) {
 }
 
 /**
+ * Resolves a eslint core config path
+ * @param {string} name The eslint config name.
+ * @returns {string} The resolved path of the config.
+ * @private
+ */
+function getEslintCoreConfigPath(name) {
+    if (name === "eslint:recommended") {
+
+       /*
+        * Add an explicit substitution for eslint:recommended to
+        * conf/eslint-recommended.js.
+        */
+        return path.resolve(__dirname, "../../conf/eslint-recommended.js");
+    }
+
+    if (name === "eslint:all") {
+
+       /*
+        * Add an explicit substitution for eslint:all to conf/eslint-all.js
+        */
+        return path.resolve(__dirname, "../../conf/eslint-all.js");
+    }
+
+    throw configMissingError(name);
+}
+
+/**
  * Applies values from the "extends" field in a configuration file.
  * @param {Object} config The configuration information.
  * @param {string} filePath The file path from which the configuration information
@@ -360,43 +406,23 @@ function applyExtends(config, filePath, relativeTo) {
 
     // Make the last element in an array take the highest precedence
     config = configExtends.reduceRight((previousValue, parentPath) => {
-
-        if (parentPath === "eslint:recommended") {
-
-            /*
-             * Add an explicit substitution for eslint:recommended to
-             * conf/eslint-recommended.js.
-             */
-            parentPath = path.resolve(__dirname, "../../conf/eslint-recommended.js");
-        } else if (parentPath === "eslint:all") {
-
-            /*
-             * Add an explicit substitution for eslint:all to conf/eslint-all.js
-             */
-            parentPath = path.resolve(__dirname, "../../conf/eslint-all.js");
-        } else if (isFilePath(parentPath)) {
-
-            /*
-             * If the `extends` path is relative, use the directory of the current configuration
-             * file as the reference point. Otherwise, use as-is.
-             */
-            parentPath = (!path.isAbsolute(parentPath)
-                ? path.join(relativeTo || path.dirname(filePath), parentPath)
-                : parentPath
-            );
-        }
-
         try {
+            if (parentPath.startsWith("eslint:")) {
+                parentPath = getEslintCoreConfigPath(parentPath);
+            } else if (isFilePath(parentPath)) {
+
+                /*
+                 * If the `extends` path is relative, use the directory of the current configuration
+                 * file as the reference point. Otherwise, use as-is.
+                 */
+                parentPath = (path.isAbsolute(parentPath)
+                    ? parentPath
+                    : path.join(relativeTo || path.dirname(filePath), parentPath)
+                );
+            }
             debug(`Loading ${parentPath}`);
             return ConfigOps.merge(load(parentPath, false, relativeTo), previousValue);
         } catch (e) {
-            if (parentPath.indexOf("plugin:") === 0 || parentPath.indexOf("eslint:") === 0) {
-                e.message = `Failed to load config "${parentPath}" to extend from.`;
-                e.messageTemplate = "extend-config-missing";
-                e.messageData = {
-                    configName: parentPath
-                };
-            }
 
             /*
              * If the file referenced by `extends` failed to load, add the path
@@ -462,7 +488,10 @@ function normalizePackageName(name, prefix) {
  * or package name.
  * @param {string} filePath The filepath to resolve.
  * @param {string} [relativeTo] The path to resolve relative to.
- * @returns {Object} A path that can be used directly to load the configuration.
+ * @returns {Object} An object containing 3 properties:
+ * - 'filePath' (required) the resolved path that can be used directly to load the configuration.
+ * - 'configName' the name of the configuration inside the plugin.
+ * - 'configFullName' the name of the configuration as used in the eslint config (e.g. 'plugin:node/recommended').
  * @private
  */
 function resolve(filePath, relativeTo) {
@@ -471,14 +500,15 @@ function resolve(filePath, relativeTo) {
     }
     let normalizedPackageName;
 
-    if (filePath.indexOf("plugin:") === 0) {
-        const packagePath = filePath.substr(7, filePath.lastIndexOf("/") - 7);
+    if (filePath.startsWith("plugin:")) {
+        const configFullName = filePath;
+        const pluginName = filePath.substr(7, filePath.lastIndexOf("/") - 7);
         const configName = filePath.substr(filePath.lastIndexOf("/") + 1, filePath.length - filePath.lastIndexOf("/") - 1);
 
-        normalizedPackageName = normalizePackageName(packagePath, "eslint-plugin");
+        normalizedPackageName = normalizePackageName(pluginName, "eslint-plugin");
         debug(`Attempting to resolve ${normalizedPackageName}`);
         filePath = resolver.resolve(normalizedPackageName, getLookupPath(relativeTo));
-        return { filePath, configName };
+        return { filePath, configName, configFullName };
     }
     normalizedPackageName = normalizePackageName(filePath, "eslint-config");
     debug(`Attempting to resolve ${normalizedPackageName}`);

--- a/tests/lib/config/config-file.js
+++ b/tests/lib/config/config-file.js
@@ -233,20 +233,63 @@ describe("ConfigFile", () => {
 
         });
 
-        it("should throw an error when a plugin config is not found", () => {
+        it("should throw an error when a parser in a plugin config is not found", () => {
+            const resolvedPath = path.resolve(PROJECT_PATH, "./node_modules/eslint-plugin-test/index.js");
 
             const configDeps = {
-                "../util/module-resolver": createStubModuleResolver({})
+                "../util/module-resolver": createStubModuleResolver({
+                    "eslint-plugin-test": resolvedPath
+                }),
+                "require-uncached"(filename) {
+                    return configDeps[filename];
+                }
+            };
+
+            configDeps[resolvedPath] = {
+                configs: {
+                    bar: {
+                        parser: "babel-eslint"
+                    }
+                }
             };
 
             const StubbedConfigFile = proxyquire("../../../lib/config/config-file", configDeps);
 
             assert.throws(() => {
                 StubbedConfigFile.applyExtends({
-                    extends: "plugin:foo",
+                    extends: "plugin:test/bar",
                     rules: { eqeqeq: 2 }
                 }, "/whatever");
-            }, /Failed to load config "plugin:foo" to extend from./);
+            }, /Cannot find module 'babel-eslint'/);
+
+        });
+
+        it("should throw an error when a plugin config is not found", () => {
+            const resolvedPath = path.resolve(PROJECT_PATH, "./node_modules/eslint-plugin-test/index.js");
+
+            const configDeps = {
+                "../util/module-resolver": createStubModuleResolver({
+                    "eslint-plugin-test": resolvedPath
+                }),
+                "require-uncached"(filename) {
+                    return configDeps[filename];
+                }
+            };
+
+            configDeps[resolvedPath] = {
+                configs: {
+                    baz: {}
+                }
+            };
+
+            const StubbedConfigFile = proxyquire("../../../lib/config/config-file", configDeps);
+
+            assert.throws(() => {
+                StubbedConfigFile.applyExtends({
+                    extends: "plugin:test/bar",
+                    rules: { eqeqeq: 2 }
+                }, "/whatever");
+            }, /Failed to load config "plugin:test\/bar" to extend from./);
 
         });
 

